lib/common/serverutil.c

```diff
@@ -259,12 +261,16 @@ int h2o_read_command(const char *cmd, char **argv, h2o_buffer_t **resp, int *chi
         goto Exit;
 
     /* spawn */
-    int mapped_fds[] = {respfds[1], 1, /* stdout of the child process is read from the pipe */
-                        -1};
-    if ((pid = h2o_spawnp(cmd, argv, mapped_fds, 1)) == -1)
+    h2o_priv_init_exec_context(&h2o_ec);
+    h2o_priv_bind_fd(&h2o_ec, respfds[1], STDOUT_FILENO, H2O_FDMAP_READ_FROM_PROC);
+    pid = h2o_priv_exec(&h2o_ec, cmd, argv, SANDBOX_POLICY_NONE);
+    if (pid == -1) {
+        h2o_priv_cleanup_exec_context(&h2o_ec);
```

Comment:
I need to go back and read the code. I think I initially went down the path of working on `h2o_spawnp` but I ran into problems with the file descriptor mappings. The file descriptor mapping and "plumbing" required different handling in the privsep case, and the initialization of the FD maps happened before the invocation of h2o_spawnp. I have since simplified some of the file descriptor mapping logic in the privsep mode, so it may be possible to add the privsep shims there but I think I would need to do some testing

Reply:
It would be really great if you could make sure that `h2o_spawnp` works. I do not want to see privilege separation suddenly breaking when we add new use of `h2o_spawnp` or when we change one of existing uses.

Reply:
I think if we change the function to accept the `h2o_fd_mapping` that is previously populated, the effectively will encode the fd mapping instructions. Let me take this away and see how it works.